File: Leetcode/0096-unique-binary-search-trees/solution.java
class Solution {
    public int numTrees(int n) {
        int[] g = new int[n + 1];
        g[0] = 1;
        if (n >= 1) g[1] = 1;
        for (int i = 2; i <= n; i++)
            for (int j = 0; j < i; j++) g[i] += g[j] * g[i - 1 - j];
        return g[n];
    }

    static int passed = 0, failed = 0;
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[][] cases = {{1, 1}, {2, 2}, {3, 5}, {4, 14}, {5, 42}, {6, 132}, {7, 429}, {19, 1767263190}};
        for (int[] c : cases) {
            int got = sol.numTrees(c[0]);
            if (got == c[1]) { System.out.println("PASS: n=" + c[0]); passed++; }
            else { System.out.println("FAIL: n=" + c[0] + " got=" + got); failed++; }
        }
        System.out.printf("%nResults: %d passed, %d failed, %d total%n", passed, failed, passed + failed);
    }
}

File: Leetcode/0096-unique-binary-search-trees/solution.py
class Solution:
    def numTrees(self, n: int) -> int:
        """Time O(n^2), Space O(n)."""
        g = [0] * (n + 1)
        g[0] = 1
        if n >= 1: g[1] = 1
        for i in range(2, n + 1):
            for j in range(i):
                g[i] += g[j] * g[i - 1 - j]
        return g[n]


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    cases = [(1, 1), (2, 2), (3, 5), (4, 14), (5, 42), (6, 132), (7, 429), (19, 1767263190)]
    for n, exp in cases:
        got = sol.numTrees(n)
        if got == exp: print(f"PASS: n={n}"); passed += 1
        else: print(f"FAIL: n={n} got={got}"); failed += 1
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
